freeipa/src/main/java/com/sequenceiq/freeipa/service/user/UsersyncService.java

```diff
@@ -0,0 +1,93 @@
+import com.sequenceiq.freeipa.client.FreeIpaClient;
+import com.sequenceiq.freeipa.client.FreeIpaClientBuilder;
+import com.sequenceiq.freeipa.client.model.RPCResponse;
+import com.sequenceiq.freeipa.entity.FreeIpa;
+import com.sequenceiq.freeipa.entity.Stack;
+import com.sequenceiq.freeipa.service.FreeIpaService;
+import com.sequenceiq.freeipa.service.GatewayConfigService;
+import com.sequenceiq.freeipa.service.TlsSecurityService;
+import com.sequenceiq.freeipa.service.stack.StackService;
+
+@Service
+public class UsersyncService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(UsersyncService.class);
+
+    private static final String STACK_NOT_FOUND_ERROR = "Stack not found for environment %s and name %s";
+
+    private static final String FREEIPA_NOT_FOUND_ERROR = "FreeIPA not found for stack %d";
+
+    @Inject
+    private GatewayConfigService gatewayConfigService;
+
+    @Inject
+    private StackService stackService;
+
+    @Inject
+    private FreeIpaService freeIpaService;
+
+    @Inject
+    private TlsSecurityService tlsSecurityService;
+
+    public void synchronizeUsers(SynchronizeUsersRequest request) {
+        LOGGER.info("UsersyncService.synchronizeUsers() called");
+
+        Stack stack = stackService.getByNameAndEnvironment(request.getName(), request.getEnvironmentName());
+        FreeIpa freeIpa = freeIpaService.findByStack(stack);
+
+        GatewayConfig primaryGatewayConfig = gatewayConfigService.getPrimaryGatewayConfig(stack);
+        HttpClientConfig httpClientConfig = tlsSecurityService.buildTLSClientConfigForPrimaryGateway(
+                stack.getId(), primaryGatewayConfig.getPublicAddress());
+
+        // TODO improve exception handling. e.g., group or user add will fail if group or user already exists
+        // TODO batch calls for improved performance. i.e., get all members of each group so we call groupAddMember once per group. investigate batch api
+        try {
+            FreeIpaClient freeIpaClient = new FreeIpaClientBuilder(
+                "admin", freeIpa.getAdminPassword(), freeIpa.getDomain().toUpperCase(),
+                httpClientConfig, stack.getGatewayport().toString())
+                .build();
+
+            for (Group group : request.getGroups()) {
+                LOGGER.debug("adding group {}", group.getName());
+                RPCResponse<Object> groupAdd = freeIpaClient.groupAdd(group.getName());
+                LOGGER.debug("Success: {}", groupAdd.getResult());
+            }
+
+            for (User user : request.getUsers()) {
+                String username = user.getName();
+                LOGGER.debug("adding user {}", username);
+                RPCResponse<com.sequenceiq.freeipa.client.model.User> userAdd = freeIpaClient.userAdd(
+                        username, user.getFirstName(), user.getLastName(), user.getPassword());
+                LOGGER.debug("Success: {}", userAdd.getResult());
+
+                // Add user to groups
```

Comment:
we shouldn't have comments like these, move this part into a method called sth like addUserToGroups